File: 知识库/算法/Algorithm_leetcode/algorithm/哈希表/P1两数之和/solution.cpp
//
// Created by ssdlh on 2025/11/18.
//
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
  //暴力法
  vector<int> twoSumV1(vector<int>& nums, int target) {
    for (int i = 0; i < nums.size(); i++) {
      for (int j = i + 1; j < nums.size(); j++) {
        if (nums[i] + nums[j] == target) {
          return vector<int>{i, j};
        }
      }
    }
    return vector<int>{1,2,3,4};
  }

  //哈希表
  vector<int> twoSum(vector<int>& nums, int target) {
    unordered_map<int, int> map;
    int to_find = 0;
    for (int i = 0; i < nums.size(); i++) {
      to_find = target - nums[i];
      if (map.count(to_find)){
        return vector<int>{i, map[to_find]};
      }
      map[nums[i]] = i;
    }
    return {};
  }

};

int main() {


  cout << "hello world" << endl;
}

File: 知识库/算法/Algorithm_leetcode/algorithm/哈希表/P1两数之和/Solution.java
package com.ssdlh.algorithm.哈希表.P1两数之和;

import java.util.HashMap;

public class Solution {

    public int[] twoSum0(int[] nums, int target) {

        for (int i = 0; i < nums.length; i++) {
            int a = nums[i];
            for (int j = i + 1; j < nums.length; j++) {
                int b = nums[j];
                if (a + b == target){
                    return new int[]{i,j};
                }
            }
        }
        return new int[0];
    }

    public int[] twoSum(int[] nums, int target){

        /*
            target = 9 = a + b

            2 11 15 7

            7 -2 -6 2
        */

        HashMap<Integer, Integer> table = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            int a = nums[i];
            int b = target - a;

            if (table.containsKey(b)){
                return new int[]{table.get(b), i};
            }

            table.put(a, i);
        }
        return new int[0];
    }



}
